Triangle numerical integration

hdeling opened the pull request:
This pull-request adds the following features 
- add numerical integration on tetrahedra and triangles
- add a mechanical tag on a PointSetGeometry object in order to find its associated MechanicalObject
- add the orthogonal of a 2d vector
- add support of high order elements in VTK and Gmesh mesh loaders

It also fixes some bugs 
-Quand2Triangle Topologivcal mapping



______________________________________________________
<!--- Please leave this at the end of your message -->
This PR: 
- [x] builds with SUCCESS for all platforms on the CI.
- [x] does not generate new warnings nor unit test failures.
- [x] does not break existing scenes.
- [x] does not break API compatibility.
- [x] has been reviewed and agreed to be transitional.
- [x] is more than 1 week old.

**Reviewers will merge only if all these checks are true.**

hdeling committed:
ADD : orthogonal of a 2d vector

hdeling committed:
ADD : test for numerical integration on triangles

hdeling committed:
FIX : bug which creates twice many vertices

hdeling committed:
ADD : tag to find the propoer mechanical object

hdeling committed:
ADD : numerical integration on triangles

hdeling committed:
ADD : add the ability to generate a 2D grid with quads and triangles. Changed the name of parameters

hdeling committed:
ADD : numerical integral on triangle of degree 8

hdeling committed:
ADD : numerical integration on simplex up to degree 10

hdeling committed:
UPDATE : added support of high order triangle and tetrahedron meshes in mesh loader

hdeling committed:
ADD : support of quadratic triangle and tetrahedron in VTK and Gmsh files

ChristianDuriez commented on SofaKernel/framework/sofa/core/loader/MeshLoader.cpp:
```diff
@@ -39,10 +39,15 @@ MeshLoader::MeshLoader() : BaseLoader()
     , d_triangles(initData(&d_triangles,"triangles","Triangles of the mesh loaded"))
     , d_quads(initData(&d_quads,"quads","Quads of the mesh loaded"))
     , d_polygons(initData(&d_polygons,"polygons","Polygons of the mesh loaded"))
+    , highOrderEdgePositions(initData(&highOrderEdgePositions,"highOrderEdgePositions","High order edge points of the mesh loaded"))
```
Hi Hervé !
Now we have naming conventions in SOFA:
https://github.com/sofa-framework/sofa/blob/master/GUIDELINES.md
see Rule N5 in Naming section: d_ for data

ChristianDuriez commented on SofaKernel/framework/sofa/core/loader/MeshLoader.h:
```diff
@@ -93,12 +111,17 @@ class SOFA_CORE_API MeshLoader : public BaseLoader
     Data< helper::vector< Triangle > > d_triangles;
     Data< helper::vector< Quad > > d_quads;
     Data< helper::vector< helper::vector <unsigned int> > > d_polygons;
+    Data< helper::vector< HighOrderEdgePosition > > highOrderEdgePositions;
```
Here also, naming data convention d_

ChristianDuriez commented on SofaKernel/modules/SofaBaseTopology/PointSetGeometryAlgorithms.h:
```diff
@@ -132,6 +128,8 @@ class PointSetGeometryAlgorithms : public core::topology::GeometryAlgorithms
     Data<float> showIndicesScale;
     Data<bool> showPointIndices;
+    /// Tage of the Mechanical State associated with the vertex position
+    Data<std::string> m_tagMechanics;
```
d_tagMechanics;

m_ is for other variable names [N7]

damienmarchal commented on modules/SofaTopologyMapping/Quad2TriangleTopologicalMapping.cpp:
```diff
@@ -99,10 +99,10 @@ void Quad2TriangleTopologicalMapping::init()
-            {
-                toModel->addPoint(fromModel->getPX(i), fromModel->getPY(i), fromModel->getPZ(i));
-            }
+            //for (int i=0; i<fromModel->getNbPoints(); i++)
```
Shouldn't the commented code removed ?

hdeling replied:
Yes of course.

ChristianDuriez commented on SofaKernel/modules/SofaBaseTopology/PointSetGeometryAlgorithms.h:
```diff
@@ -132,6 +128,8 @@ class PointSetGeometryAlgorithms : public core::topology::GeometryAlgorithms
     sofa::core::topology::BaseMeshTopology* m_topology;
     Data<float> showIndicesScale;
```
These data could also be renamed

damienmarchal commented on SofaKernel/modules/SofaBaseTopology/SofaBaseTopology_test/TriangleNumericalIntegration_test.cpp:
```diff
@@ -0,0 +1,165 @@
+		Vec<3,unsigned short> randomPolynomial;
+
+		// get the descriptor of numerical integration on tetrahedra
+        NumericalIntegrationDescriptor &nid=geo->getTriangleNumericalIntegrationDescriptor();
```
Identation looks weird.

damienmarchal commented:
Hi @hdeling,

Thank you for your PR. 
At first glance it looks nice only cosmetic/minor details.

ChristianDuriez commented on modules/SofaGeneralEngine/GenerateGrid.inl:
```diff
@@ -54,16 +52,14 @@ GenerateGrid<DataTypes>::GenerateGrid()
-    addInput(&f_resolutionLength);
-    addInput(&f_resolutionWidth);
-    addInput(&f_resolutionHeight);
+    addInput(&f_minCorner);
```
If you can change all the data from
f_ 
to 
d_ 
it would be great ;-))

ChristianDuriez reviewed (changes requested):
Thanks Hervé ! 
Very useful changes.  I didn't really get if it can be used for computation of internal force in Triangle or Tetrahedron ForceFields ?

Only the same change requested
put d_  before datatypes 
(see N5 at https://github.com/sofa-framework/sofa/blob/master/GUIDELINES.md)

hdeling committed:
ADD : Ignore files

hdeling committed:
UPDATE : indentation
FIX : replaced sofa::helper::set
REMOVE : commented lines
UPDATE : renamed data members with prefix d_

guparan committed:
[all] FIX indentation

guparan committed:
[SofaGeneralEngine] FIX Data declaration order

To avoid warnings on build.

guparan commented:
Should we revert .gitignore modification in cfa5a44 ?

hdeling commented:
Yes probably should remove the commit on .gitignore.

hdeling committed:
UPDATE : reset .gitignore

BenjaminGilles reviewed (approved):
thanks Hervé, this looks all good to me.
minor comment:  some functions might be moved from  SofaBaseTopology/CommonAlgorithms.h  to defaulttype/Vec.h

damienmarchal commented:
[ci-build]

damienmarchal committed:
Merge branch 'master' into TriangleNumericalIntegration

damienmarchal committed:
Update CMakeLists.txt

BezierTetrahedronTopology_test was removed.

damienmarchal commented:
Hi all,

It seems there a lot of new test-scene failures since the merge of this PR. 
https://ci.inria.fr/sofa-ci/job/centos_clang-3.4_options/635/console

We probably need to fix that in a new PR.
